popmon/visualization/utils.py

```diff
@@ -46,6 +47,154 @@ def xtick(lab, top):
+
+    if td.days % 7 > 0:
+        time_delta = time_delta + " " + str(td.days % 7) + "d"
+
+    if td.seconds // 3600 > 0:
+        time_delta = time_delta + " " + str(td.seconds // 3600) + "h"
+
+    if (td.seconds // 60) % 60 > 0:
+        time_delta = time_delta + " " + str((td.seconds // 60) % 60) + "m"
+
+    if td.seconds % 60 > 0:
+        time_delta = time_delta + " " + str(td.seconds % 60) + "s"
+
+    return time_delta
+
+
+def get_summary_table(
+    num_features: int,
+    time_bins: int,
+    time_axis: str,
+    reference_type: str,
+    time_width: float,
+    offset: str,
+    max_timestamp: str,
+) -> dict:
+    """
+    This function returns a dictionary of summary statistics for a given set of features
+
+    :param num_features: number of features in the dataset
+    :param time_bins: number of time bins to use
+    :param time_axis: the name of the time axis
+    :param reference_type: the type of reference to use
+    :param time_width: the width of the time bins in seconds
+    :param offset: the time offset from the start of the data
+    :param max_timestamp: the maximum timestamp in the data
+    :rtype: dict
+    """
+
+    bin_width = convert_time_delta(time_width)
+    data = {
+        "time axis": time_axis,
+        "reference type": reference_type,
+        "number of features": num_features,
+        "time bins": time_bins,
+        "bin width": bin_width if len(bin_width) > 0 else time_width,
+        "time range": str(offset) + " - " + str(max_timestamp),
+    }
+
+    # plot summary table
+    fig = go.Figure(
+        data=[
+            go.Table(
+                header={
+                    "fill_color": "white",
+                    "line": {"color": "white"},
+                },
+                cells={
+                    "values": [list(data.keys()), list(data.values())],
+                    "fill": {"color": ["rgb(235, 240, 248)", "rgb(235, 240, 248)"]},
+                    "line": {"color": "white"},
+                    "align": "left",
+                },
+            )
+        ]
+    )
+
+    fig.update_layout(
+        autosize=False,
+        margin={"l": 10, "r": 10, "t": 10, "b": 10},
+        height=200,
+        width=450,
+    )
+
+    plot = json.loads(fig.to_json())
+
+    return {
+        "name": "Dataset Summary",
+        "type": "overview",
+        "description": "",
+        "plot": plot.get("data", ""),
+        "layout": plot.get("layout", ""),
+    }
+
+
+def get_reproduction_table(start_time: str, version: str):
+    """
+    > This function returns a table of the reproduction data for a given version of the model
+
+    :param start_time: The time at which the experiment was started
+    :type start_time: str
+    :param version: the version of the model you want to use
+    :type version: str
+    """
+
+    data = {
+        "time started": start_time,
+        "time ended": datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
```

Comment:
This time is a bit arbitrary (in the middle of the report generation).
A cleaner way of achieving the same is adding a `TimingModule` to the pipelines.
The module would take a `key` that stores the `datetime.now()` to that key in the datastore. This way it's reusable for start- and end time in the pipeline framework.

For end time, it can be placed just before ReportGenerator.